### PR TITLE
Migrate tests to `nose2` instead of `nose`

### LINKED ISSUE #746: Migrate from `nose` to `nose2` for testing
`nose` no longer works with Python 3.11 so we need to move to `nose2` if we want to support Python 3.11

### PR DESCRIPTION
**Pull request by desilinguist**
- Update all tests to be nose2-compatible. 
- Update pre-commit config to ignore docstyle and flake8 errors in test files. 
- Replace `nose` with `nose2` in requirements file.
- Update Gitlab and Azure CI builds to use nose2. 
- Add `unittest.cfg` for nose2 configuration 
- Move testing utilities into new `skll.utils.testing` module since `nose2` doesn't pick up the `tests` directory as a module. 
- Update all tests to use `skll.utils.testing`. 
- Update `.coveragerc` with proper exclusions. 
- Disable logging in very noisy test file.

This PR closes #746.

### CODE REVIEW

**Comment by mulhod** [comment]
There are still references to `nosetests` in `CONTRIBUTING.md`. Also it is said there that `python=3.10` is required for setting up a development environment. Maybe that's still true for other reasons, but I think it should be considered. I actually set up an environment that uses 3.11 and I think the unit tests are running fine with `nose2`.

**Comment by desilinguist** [comment]
> There are still references to `nosetests` in `CONTRIBUTING.md`. Also it is said there that `python=3.10` is required for setting up a development environment. Maybe that's still true for other reasons, but I think it should be considered. I actually set up an environment that uses 3.11 and I think the unit tests are running fine with `nose2`.

Good catch on `CONTRIBUTING.md`. As for Python 3.11, there's a separate issue #744 for that which I will work on after this one is merged. As part of that, I will be switching the Azure build to 3.11 from 3.9.

**Comment by mulhod** [comment]
All unit tests passed with 3.11.

**Review by mulhod — APPROVED** [review]

File: skll/utils/testing.py
Comment (dblandan) [inline skll/utils/testing.py:1]: I noticed that the methods with returns don't have type hints for the return values...  I'm not 100% sure, but I think the other recent PRs included these.  (e.g., the `def` for `fill_in_config_paths(...)` should probably include `-> Path`.

File: tests/test_classification.py
Comment (dblandan) [inline tests/test_classification.py:130]: Not critical, but I'd prefer to see separate tests for error handling.

Comment (desilinguist) [reply tests/test_classification.py:130]: Care to elaborate?

File: tests/test_classification.py
Comment (dblandan) [inline tests/test_classification.py:1360]:
```suggestion
                # 5. The only ones left are the weighted kappas;
```

File: tests/test_commandline_utils.py
Comment (dblandan) [inline tests/test_commandline_utils.py:1]: Another general observation: I really appreciate the effort & scope of this refactor! Will there be variable type hints added later? There are a few places where I think they would be helpful, but I don't think it's worth holding up this PR for them...

Comment (desilinguist) [reply tests/test_commandline_utils.py:1]: We aren't adding type hints to test files.

File: tests/test_input.py
Comment (dblandan) [inline tests/test_input.py:885]: ok... i know this is from the pre-commit, but yuck...  lol

Comment (dblandan) [reply tests/test_input.py:885]: i feel like it might actually be more readable for these to do something like:
```
ph = parse_config_file(config_path)
do_grid_search = ph[14]
grid_objectives = ph[15]
output_metrics = ph[-2]
```

File: tests/test_input.py
Comment (dblandan) [inline tests/test_input.py:959]: Especially in cases where you only need one value...
```suggestion
        train_path = parse_config_file(config_path)[37]
```

**Review by dblandan — APPROVED** [review]

File: tests/test_input.py
Comment (desilinguist) [reply tests/test_input.py:885]: Good call! Definitely less yucky :)

**Pull request merged**
